src/common/config.h

```diff
@@ -2,84 +2,161 @@
  * Copyright 2014-2019 by Contributors
  * \file config.h
  * \brief helper class to load in configures from file
- * \author Haoda Fu
+ * \author Haoda Fu, Hyunsu Cho
  */
 #ifndef XGBOOST_COMMON_CONFIG_H_
 #define XGBOOST_COMMON_CONFIG_H_
 
+#include <dmlc/logging.h>
 #include <cstdio>
-#include <cstring>
 #include <string>
 #include <istream>
 #include <fstream>
 #include <vector>
+#include <regex>
+#include <iterator>
 #include <utility>
 
 namespace xgboost {
 namespace common {
 /*!
  * \brief Implementation of config reader
  */
-class ConfigParse {
+class ConfigParser {
  public:
   /*!
-   * \brief constructor
-   * \param cfgFileName name of configure file
+   * \brief Constructor for INI-style configuration parser
+   * \param path path to configuration file
    */
-  explicit ConfigParse(const std::string &cfgFileName) {
-    fi_.open(cfgFileName);
-    if (fi_.fail()) {
-      LOG(FATAL) << "cannot open file " << cfgFileName;
-    }
+  explicit ConfigParser(const std::string& path)
+    : line_comment_regex_("^#"),
+      key_regex_(R"rx(^([^#"'=\r\n\t ]+)[\t ]*=)rx"),
+      key_regex_escaped_(R"rx(^(["'])([^"'=\r\n]+)\1[\t ]*=)rx"),
+      value_regex_(R"rx(^([^#"'=\r\n\t ]+)[\t ]*(?:#.*){0,1}$)rx"),
```

Comment:
the value payload may also contain token `/` `-` etc - `hdfs://` `s3://` `.fmt-` etc. they may need to be handled as well

Reply:
These letters are fine, because in the value, all letters except `#`, `"`, `'`, `=`, `\r`, `\n`, `\t`, and ` ` (space) are allowed. (`#`, `\t`, and ` ` may be inside the value if the value is quoted)